### DISCUSSION TITLE
Allow custom store implementations to not inherit from rdflib.store.Store

### DISCUSSION THREAD

**Discussion by ashleysommer** [RDFLib/rdflib#1538] [Ideas]
Opening this as a discussion topic.

Currently, when instantiating a `Graph` instance with a pass-in `store` instance, the `store` parameter must be either:
1) An instance of something that is a subclass of `rdflib.store.Store`, _or_
2) A string, to dereference a store implementation from a registered plugin

See:
https://github.com/RDFLib/rdflib/blob/ab31c5ed3772cb63806a4614ca4431e9bd7de8d3/rdflib/graph.py#L325-L329

However, I've twice encountered the situation where I want to pass in a store that looks like and acts like a `rdflib.store.Store` (via duck-typing) but does not inherit from `rdflib.store.Store`.

Eg, this doesn't work:
```python
from nonrdflibmodule.store import MySpecialStore
from rdflib import Graph
mystore = MySpecialStore()
g = Graph(store=mystore)  #<- fails
```
This situation, the Graph constructor sees that `MySpecialStore` does not inherit from `rdflib.store.Store` so it uses that as a lookup key to find a plugin, and fails because it is not a valid lookup key.

There is a workaround I've found, by temporarily registering the store:
```python
from rdflib.plugin import register
register("tempstore", Store, "nonrdflibmodule.store", "MySpecialStore")
g = Graph(store="tempstore")
mystore = g.store
```

I know this is a niche situation, but I wonder if we could use some kind of duck-typing mechanism to check if a store implements a basic set of features, and allow it, even if it not a subclass of `rdflib.store.Store`.

**Comment by aucampia** [comment]
Maybe the best option would be to just invert this logic:

https://github.com/RDFLib/rdflib/blob/ab31c5ed3772cb63806a4614ca4431e9bd7de8d3/rdflib/graph.py#L325-L329

i.e., if it is a string, lookup the plugin, if not, just proceed as is. There is also https://docs.python.org/3/library/typing.html#typing.runtime_checkable - but that would be best used when no other option is available/appropriate. Something like `runtime_checkable` can be done by just selecting some methods to check for that should exist on `Store` but not on `str`. I think in general duck-typing should be supported wherever possible.

**Comment by nicholascar** [comment]
I don’t think the just-in-time registration’s a workaround: that’s the expected solution, as per other parsers and serialisers’   registration.
